modules/kargoBidAdapter.js

```diff
@@ -28,57 +28,96 @@ export const spec = {
   },
   buildRequests: function(validBidRequests, bidderRequest) {
     const currencyObj = config.getConfig('currency');
-    const currency = (currencyObj && currencyObj.adServerCurrency) || 'USD';
-    const bidIDs = {};
-    const bidSizes = {};
+    const currency = (currencyObj && currencyObj.adServerCurrency) ? currencyObj.adServerCurrency : null;
+    const impressions = [];
 
     _each(validBidRequests, bid => {
-      bidIDs[bid.bidId] = bid.params.placementId;
-      bidSizes[bid.bidId] = bid.sizes;
+      impressions.push(spec.getImpression(bid))
     });
 
     const firstBidRequest = validBidRequests[0];
 
     const tdid = deepAccess(firstBidRequest, 'userId.tdid')
 
-    const transformedParams = Object.assign({}, {
-      sessionId: spec._getSessionId(),
-      requestCount: spec._getRequestCount(),
+    const krakenParams = Object.assign({}, {
+      pbv: '$prebid.version$',
+      aid: firstBidRequest.auctionId,
+      sid: spec._getSessionId(),
+      url: spec._getAllMetadata(bidderRequest).pageURL,
       timeout: bidderRequest.timeout,
-      currency,
-      cpmGranularity: 1,
-      timestamp: (new Date()).getTime(),
-      cpmRange: {
-        floor: 0,
-        ceil: 20
-      },
-      bidIDs,
-      bidSizes,
+      ts: new Date().getTime(),
       device: {
-        width: window.screen.width,
-        height: window.screen.height
+        size: [
+          window.screen.width,
+          window.screen.height
+        ]
       },
-      prebidRawBidRequests: validBidRequests
-    }, spec._getAllMetadata(bidderRequest, tdid));
+      imp: impressions,
+      user: spec._getUserIds(tdid, bidderRequest.uspConsent, bidderRequest.gdprConsent),
+      eids: firstBidRequest.userIdAsEids
+    });
+
+    const reqCount = spec._getRequestCount()
+    if (reqCount != null && reqCount > 0) {
+      krakenParams.requestCount = reqCount
+    }
+
+    if (currency != null && currency != 'USD') {
+      krakenParams.cur = currency
+    }
+
+    // Pull Social Canvas segments and embed URL
+    const socialCanvas = deepAccess(firstBidRequest, 'params.socialCanvas');
+
+    if (socialCanvas != null) {
+      krakenParams.socan = socialCanvas
+    }
 
     // User Agent Client Hints / SUA
     const uaClientHints = deepAccess(firstBidRequest, 'ortb2.device.sua');
     if (uaClientHints) {
-      transformedParams.device.sua = pick(uaClientHints, ['browsers', 'platform', 'mobile', 'model']);
+      const suaAttributes = ['browsers', 'platform', 'mobile', 'model', 'source']
+      let suaValidAttributes = []
+
+      suaAttributes.forEach(attributeKey => {
```

Comment:
thoughts on doing something like this to improve readability?
```
      for (const suaKey of suaAttributes) {
        const suaValue = uaClientHints[suaKey]
        if (!suaValue) {
          continue
        }
        if (typeof suaValue !== 'string') {
          continue
        }
        if (suaValue.trim() === '') {
          continue
        }

        switch (suaKey) {
          case 'mobile', 'source':
            if (suaValue.length < 1) {
              return;
            };
          default:
            suaValidAttributes.push(suaKey);
        }
      }
```

Reply:
Implemented a slightly modified version of this as not all values are strings.